pypsa/components.py

```diff
@@ -440,9 +448,45 @@ def snapshot_weightings(self, df):
             self._snapshot_weightings = df
 
 
+    @property
+    def investment_period_weightings(self):
+        """
+        Weightings applied to each investment period during the optimization
+        (LOPF).
+
+        * Objective weightings are multiplied with all cost coefficients in the
+          objective function of the respective investment period
+          (e.g. to include a social discount rate).
+
+        * Time weightings denote the elapsed time until the subsequent investment period
+          (e.g. used for global constraints CO2 emissions).
+
+        """
+        return self._investment_period_weightings
+
+
+    @investment_period_weightings.setter
+    def investment_period_weightings(self, df):
+        # if multiindex snapshots
+        if isinstance(self.snapshots, pd.MultiIndex):
+            assert df.index.equals(self.snapshots.levels[0]), "Weightings not defined for all investment periods."
+            assert (all(isinstance(x, int) for x in df.index)
+                    and all(sorted(df.get_level_values(0).unique()) == df.index.get_level_values(0).unique())), "Investment periods should be integer and increasing."
+            if isinstance(df, pd.Series):
+                logger.info('Applying weightings to all columns of `investment_period_weightings`')
+                df = pd.DataFrame({c: df for c in self._investment_period_weightings.columns})
+                self._snapshot_weightings = df
+            else:
+                self._investment_period_weightings = df
+        # for single optimisation return df
+        else:
+            self._investment_period_weightings = df
+
+
```

Comment:
The idea here is good, but the implementation mixes stuff. 
I think we need the in-between layer with `n.investment_periods` here. It would be cleaner that investment periods are only set in a function `set_investment_periods` and investment weightings only in this functions. So setting investment period weightings must not have any effect on snapshot weightings and investment periods.